python/cudf/cudf/core/column/categorical.py

```diff
@@ -261,14 +255,8 @@ def _fill(
         return result
 
     def slice(self, start: int, stop: int, stride: int | None = None) -> Self:
-        codes = self.codes.slice(start, stop, stride)
-        return type(self)(
-            data=self.data,  # type: ignore[arg-type]
-            size=codes.size,
-            dtype=self.dtype,
-            mask=codes.base_mask,
-            offset=codes.offset,
-            children=(codes,),
+        return self.codes.slice(start, stop, stride)._with_type_metadata(  # type: ignore[return-value]
```

Comment:
What is the actual return type here? IIUC, `self.codes` is a `Series`? But we say we return `Self`, a `CategoricalColumn`.

Reply:
`self.codes` is a `NumericalColumn`. 

Unfortunately the typing of `_with_type_metadata` isn't entirely correct everywhere yet. In this case `NumericalColumn._with_type_metadata` can return `Self | CategoricalColumn`. Here I'd like to express that `CategoricalColumn.slice` returns `Self` since `NumericalColumn._with_type_metadata` will return `CategoricalColumn`. So I probably would need to use a `cast` for that?